File: go/pkg/vecstore/vecstore.go
// Package vecstore provides a vector approximate nearest-neighbor (ANN)
// search interface and implementations.
//
// The [Index] interface defines the contract for vector storage and search.
// Implementations include an in-memory brute-force index for testing
// ([NewMemory]) and HNSW for production use (planned).
//
// This package follows the same pattern as [kv]: a generic interface with
// pluggable backends. For all-in-one deployment, use the built-in
// implementations. For distributed deployment, swap in a client that
// talks to Milvus, Qdrant, or similar.
package vecstore

// Index is the interface for approximate nearest-neighbor search over
// dense float32 vectors.
//
// All implementations must be safe for concurrent use.
type Index interface {
	// Insert adds or updates a vector with the given ID.
	Insert(id string, vector []float32) error

	// BatchInsert adds or updates multiple vectors at once.
	// ids and vectors must have the same length.
	BatchInsert(ids []string, vectors [][]float32) error

	// Search returns the top-k nearest vectors to the query.
	// Results are ordered by ascending distance (closest first).
	Search(query []float32, topK int) ([]Match, error)

	// Delete removes a vector by ID. No error if ID does not exist.
	Delete(id string) error

	// Len returns the number of vectors in the index.
	Len() int

	// Flush ensures all pending writes are visible to subsequent searches.
	// For in-memory implementations this is typically a no-op.
	Flush() error

	// Close releases resources held by the index.
	Close() error
}

// Match is a single result from a vector similarity search.
type Match struct {
	// ID is the identifier of the matched vector.
	ID string

	// Distance is the distance between the query and matched vector.
	// Lower values indicate higher similarity.
	Distance float32
}

File: rust/vecstore/src/vecstore.rs
use crate::error::VecError;

/// Match is a single result from a vector similarity search.
#[derive(Debug, Clone)]
pub struct Match {
    /// Identifier of the matched vector.
    pub id: String,

    /// Distance between the query and matched vector.
    /// Lower values indicate higher similarity.
    pub distance: f32,
}

/// VecIndex is the interface for approximate nearest-neighbor search over
/// dense float32 vectors.
///
/// All implementations must be safe for concurrent use (Send + Sync).
pub trait VecIndex: Send + Sync {
    /// Add or update a vector with the given ID.
    fn insert(&self, id: &str, vector: &[f32]) -> Result<(), VecError>;

    /// Add or update multiple vectors at once.
    /// `ids` and `vectors` must have the same length.
    fn batch_insert(&self, ids: &[&str], vectors: &[&[f32]]) -> Result<(), VecError>;

    /// Return the top-k nearest vectors to the query, ordered by ascending
    /// distance (closest first).
    fn search(&self, query: &[f32], top_k: usize) -> Result<Vec<Match>, VecError>;

    /// Remove a vector by ID. No error if ID does not exist.
    fn delete(&self, id: &str) -> Result<(), VecError>;

    /// Return the number of vectors in the index.
    fn len(&self) -> usize;

    /// Return true if the index contains no vectors.
    fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Ensure all pending writes are visible to subsequent searches.
    fn flush(&self) -> Result<(), VecError>;
}
